src/run_info.c

```diff
@@ -197,20 +197,26 @@ void RAI_FreeDagOp(RedisModuleCtx *ctx, RAI_DagOp *dagOp) {
     }
 }
 
-void RAI_FreeRunInfo(RedisModuleCtx *ctx, struct RedisAI_RunInfo *rinfo) {
-    if (!rinfo) {
-        return;
+long long RAI_DagRunInfoDecreaseFetch(RedisAI_RunInfo *rinfo) {
+    long long ref_count = __atomic_sub_fetch(rinfo->dagRefCount, 1, __ATOMIC_RELAXED);
+    RedisModule_Assert(ref_count >= 0);
+    if (rinfo->dagDeviceOps) {
+        array_free(rinfo->dagDeviceOps);
     }
-    if (*rinfo->dagRefCount > 0) {
-        if (rinfo->dagDeviceOps) {
-            array_free(rinfo->dagDeviceOps);
-        }
+    // If this is the last run info copy we do not free it, the OnFinish callback may free.
+    if (ref_count > 0)
         RedisModule_Free(rinfo);
+    return ref_count;
+}
+
+void RAI_FreeRunInfo(RedisModuleCtx *ctx, struct RedisAI_RunInfo *rinfo) {
+    if (!rinfo) {
         return;
-    } else {
-        pthread_rwlock_destroy(rinfo->dagLock);
-        RedisModule_Free(rinfo->dagLock);
     }
+    long long ref_count = __atomic_load_n(rinfo->dagRefCount, __ATOMIC_RELAXED);
```

Comment:
if the function is being called only when ref_count is 0, there is no need for atomic, since the caller probably already called free on the shallow copy and knows the ref count is 0